## 200968182 LAB 6 

#### Q1. Read about the movielens dataset and write down a summary of metadata.

METADATA of MOVIE DATASET

The whole dataset is in the form of comma seperated values. The data set consists of 4 csv files named links, movies,ratings and tags
  - ratings.csv - all ratings contained in the file ratings.csv
                - ratings made from 0-5 with 0.5 incrementation
  - tags.csv - contains all the tags
  - movies.csv - contains information about movie titles and the gneres associated with the movie
  - links.csv - contains identifiers that can be used to link other sources of movies data
              - movieId - identifier movieLens
              - imdbId - identifier imdb 
              - tmdbId - identifier  theMovieDb
-This data set describes a 5 - star rating and free-text tagging activity from MovieLens(a movie recommender service)
100836 ratings
3683 tag applications
9742 movies
users involed in creation = 610
duration of collection of data - March 29, 1996 - September 24, 2018
every user rated atleaset 20 ratings

#### Q2. Read the “ratings.csv” file and create a pivot table with index=‘userId’, columns=‘movieId’, values = “rating. 

In [9]:
# importing the libraries
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns

In [13]:
# reading the movies dataset
ratings = pd.read_csv(r'C:\\Pravallika\MANIPAL\THIRD SEMESTER\IDA\IDA_200968182\movielens\ratings.csv')

In [15]:
# viewing the first few rows of the movies dataset
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

To solve Question 2, we use the pivot_table() function from the Pandas library, and pass the values 'userId', 'movieId, and 'rating' to the parameters index, columns, and values respectively. This way, every individual row represents a user and columns represent movies.

In [16]:
# creating a pivot table
pd.pivot_table(ratings, index='userId', columns='movieId', values='rating')

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
...         ...     ...     ...     ...     ...     ...     ...     ...   
606         2.5     NaN     NaN     NaN     NaN     NaN     2.5     NaN   
607         4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
608         2.5     2.0     2.0     NaN     NaN     NaN     NaN     NaN   
609         3.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
610         5.0     NaN     NaN     NaN     NaN     5.0     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
606         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
607         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
608         NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
609         NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
610         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  
...         ...     ...     ...     ...  
606         NaN     NaN     NaN     NaN  
607         NaN     NaN     NaN     NaN  
608         NaN     NaN     NaN     NaN  
609         NaN     NaN     NaN     NaN  
610         NaN     NaN     NaN     NaN  

[610 rows x 9724 columns]

As we can see, there are a lot of NaN or missing values in the dataset. To handle the missing values, we will fill them with a constant. The constant has to be a value that does not clash with the values in the dataset. To do so, let's find the ratings present in the dataset.

In [17]:
# finding all the values in the rating column
ratings.rating.unique()

array([4. , 5. , 3. , 2. , 1. , 4.5, 3.5, 2.5, 0.5, 1.5])

As we can see, no user has given a rating of '0' to any movie. Hence, we use the constant 0 to fill in the missing values, and assign the pivot table to a variable names df to be used for further analysis.

In [18]:
# filling the missing values in the pivot table with 0 and assigning it to a variable
df = pd.pivot_table(ratings, index='userId', columns='movieId', values='rating', fill_value=0)
df

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     0.0     4.0     0.0     0.0     4.0     0.0       0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
5           4.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
606         2.5     0.0     0.0     0.0     0.0     0.0     2.5       0   
607         4.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
608         2.5     2.0     2.0     0.0     0.0     0.0     0.0       0   
609         3.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
610         5.0     0.0     0.0     0.0     0.0     5.0     0.0       0   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0       0       0       0     0.0       0   
2           0.0     0.0  ...     0.0       0       0       0     0.0       0   
3           0.0     0.0  ...     0.0       0       0       0     0.0       0   
4           0.0     0.0  ...     0.0       0       0       0     0.0       0   
5           0.0     0.0  ...     0.0       0       0       0     0.0       0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
606         0.0     0.0  ...     0.0       0       0       0     0.0       0   
607         0.0     0.0  ...     0.0       0       0       0     0.0       0   
608         0.0     4.0  ...     0.0       0       0       0     0.0       0   
609         0.0     4.0  ...     0.0       0       0       0     0.0       0   
610         0.0     0.0  ...     0.0       0       0       0     0.0       0   

movieId  193583  193585  193587  193609  
userId                                   
1           0.0     0.0     0.0       0  
2           0.0     0.0     0.0       0  
3           0.0     0.0     0.0       0  
4           0.0     0.0     0.0       0  
5           0.0     0.0     0.0       0  
...         ...     ...     ...     ...  
606         0.0     0.0     0.0       0  
607         0.0     0.0     0.0       0  
608         0.0     0.0     0.0       0  
609         0.0     0.0     0.0       0  
610         0.0     0.0     0.0       0  

[610 rows x 9724 columns]

As we can see, the missing values are now filled with 0. Hence, a rating more than 0 represents that the user has watched the movie and given the rating, and a value 0 represents that the user has not watched the movie. Another reason for filling the missing values with 0 was to make the pivot table numeric for further analysis.

#### Q3, Q4, Q5  Importing the related libraries

In [19]:
# importing the packages
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine, correlation

#### Q6. Find the 5 most similar user for user with user Id 25

To solve Question 6, we first compute pairwise distances between the rows, that is the users, based on the cosine similarity. To do this, we use the pairwise_distances() function from the ScikitLearn library and set the value of the parameter metric to 'cosine'.

In [21]:
# computing pairwise distances using cosine similarity between the users
distance_matrix = pairwise_distances(df, metric='cosine')

Next, we splice and select the 24th row from the resultant distance matrix to select the 25th User. We do this because arrays in Python are 0-indexed and the UserIDs start from 1. Post that, we sort the cosine similarities in the descending order to find the smallest distance between users, and hence the most similar users. We use the argsort() function on the negated array to return the indices, and hence the UserIDs of the sorted array. Then, we splice the array to find the top 5 similar users. To do this, we splice values from 1 to 6. We leave out the index 0 because it is by default the same user. After splicing, we add 1 to the indices since arrays in Python are 0-indexed and UserIDs start from 1. 

In [22]:
# finding the top 5 most similar users
(-distance_matrix[24]).argsort()[1:6]+1

array([161, 163, 546, 170, 175], dtype=int64)

As we can see, the users 161, 163, 546, 170, and 175 are most similar to user 25.

#### Q7. Use the “movies” dataset to find out the names of movies, user 1 and user 338 have watched in common and how they have rated each one of them.

In [23]:
movies=pd.read_csv("C:\\Pravallika\MANIPAL\THIRD SEMESTER\IDA\IDA_200968182\movielens\movies.csv")
movies

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

In [29]:
set(ratings.loc[ratings.userId==1, 'movieId']).intersection(set(ratings.loc[ratings.userId==338, 'movieId']))

{50, 296, 527, 593, 2959}

#### Q8. Use the movies dataset to find out the common movie names between user 2 and user 338 with least rating of 4.0

In [32]:
set(ratings.loc[ratings.userId==2, 'movieId']).intersection(set(ratings.loc[ratings.userId==338, 'movieId']))

{318, 6874}

In [33]:
print('Movie ID 318: ', movies.loc[movies.movieId==318, 'title'].values[0])
print('Rating by User 2: ', ratings.loc[((ratings.userId==2) & (ratings.movieId==318)), 'rating'].values[0])
print('Rating by User 338: ', ratings.loc[((ratings.userId==338) & (ratings.movieId==318)), 'rating'].values[0])
print()
print('Movie ID 6874: ', movies.loc[movies.movieId==6874, 'title'].values[0])
print('Rating by User 2: ', ratings.loc[((ratings.userId==2) & (ratings.movieId==6874)), 'rating'].values[0])
print('Rating by User 338: ', ratings.loc[((ratings.userId==338) & (ratings.movieId==6874)), 'rating'].values[0])

Movie ID 318:  Shawshank Redemption, The 
Rating by User 2:  3.0
Rating by User 338:  5.0

Movie ID 6874:  Kill Bill: Vol. 1 
Rating by User 2:  4.0
Rating by User 338:  4.5


In [34]:
set(ratings.loc[((ratings.userId==2) & (ratings.rating>=4.0)), 'movieId']).intersection(set(ratings.loc[((ratings.userId==338) & (ratings.rating>=4.0)), 'movieId']))

{6874}

In [35]:
print('Movie ID 6874: ', movies.loc[movies.movieId==6874, 'title'].values[0])

Movie ID 6874:  Kill Bill: Vol. 1 


#### Q9.Create a pivot table for representing the similarity among movies using correlation

In [30]:
df = pd.pivot_table(ratings, index='movieId', columns='userId', values='rating', fill_value=0)
df

userId   1    2    3    4    5    6    7    8    9    10   ...  601  602  603  \
movieId                                                    ...                  
1          4  0.0  0.0    0    4    0  4.5    0    0  0.0  ...  4.0    0    4   
2          0  0.0  0.0    0    0    4  0.0    4    0  0.0  ...  0.0    4    0   
3          4  0.0  0.0    0    0    5  0.0    0    0  0.0  ...  0.0    0    0   
4          0  0.0  0.0    0    0    3  0.0    0    0  0.0  ...  0.0    0    0   
5          0  0.0  0.0    0    0    5  0.0    0    0  0.0  ...  0.0    0    0   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
193581     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193583     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193585     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193587     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193609     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   

userId   604  605  606  607  608  609  610  
movieId                                     
1          3  4.0  2.5    4  2.5    3  5.0  
2          5  3.5  0.0    0  2.0    0  0.0  
3          0  0.0  0.0    0  2.0    0  0.0  
4          0  0.0  0.0    0  0.0    0  0.0  
5          3  0.0  0.0    0  0.0    0  0.0  
...      ...  ...  ...  ...  ...  ...  ...  
193581     0  0.0  0.0    0  0.0    0  0.0  
193583     0  0.0  0.0    0  0.0    0  0.0  
193585     0  0.0  0.0    0  0.0    0  0.0  
193587     0  0.0  0.0    0  0.0    0  0.0  
193609     0  0.0  0.0    0  0.0    0  0.0  

[9724 rows x 610 columns]

In [31]:
distance_matrix = pairwise_distances(df, metric='correlation')
distance_matrix

array([[0.        , 0.7686735 , 0.8267872 , ..., 1.0289057 , 1.0289057 ,
        1.0289057 ],
       [0.7686735 , 0.        , 0.80805513, ..., 1.01829073, 1.01829073,
        1.01829073],
       [0.8267872 , 0.80805513, 0.        , ..., 1.01172923, 1.01172923,
        1.01172923],
       ...,
       [1.0289057 , 1.01829073, 1.01172923, ..., 0.        , 0.        ,
        1.00164204],
       [1.0289057 , 1.01829073, 1.01172923, ..., 0.        , 0.        ,
        1.00164204],
       [1.0289057 , 1.01829073, 1.01172923, ..., 1.00164204, 1.00164204,
        0.        ]])

#### Q10. Find the top 5 movies which are similar to the movie “Godfather”

In [20]:
for movie in movies.title:
    if('Godfather' in str(movie)):
        print(movie)

Godfather, The (1972)
Godfather: Part II, The (1974)
Godfather: Part III, The (1990)
Tokyo Godfathers (2003)
The Godfather Trilogy: 1972-1990 (1992)
